### ISSUE TITLE
Investigate on the delay for withdrawing allocations

### ISSUE THREAD

**Issue by lubtd** [tendermint/spn#573]
We specified using a param `WithdrawalAllocationdelay` for the delay when allocations for an auction can be withdrawn (after the auction is finished). In our current design, a user may undelegate stake and delegate it with another account for having new allocations, so it may make sense to set the allocation withdrawal delay to the unbonding period independently on if the auction is finished.

This task is about doing the investigation on what is the best way to represent the delay before allocations for an auction can be withdrawn and implementing the structure to be able to determine the delay (like for example adding the `WithdrawalAllocationdelay`)

Depends on https://github.com/tendermint/spn/issues/569

**Comment by giunatale** [comment]
*TLDR;* I propose we remove the `WithdrawalAllocationdelay` and just prevent withdrawal until the auction has **started**, not *ended*.
What should be done instead is add a check in `msgParticipate` to prevent participation if the auction start time is further away than the unbonding period.


### Long version

The keeper function to get total allocations already provides a number that excludes unbonding delegations. 

Therefore, assuming an account has:

1. bonded some tokens
2. used the resulting allocations to participate to an auction (this can only be done before the auction has started with current design)

he then could:

3. unbond right after sending `msgParticipate`
4. transfer funds to another wallet, and 
5. rinse and repeat...

This is worthwile doing actually **before** an auction has started, and not **after** (since again, you can't participate anymore).
Withdrawing an allocation from an auction would only make usable again **valid** allocations (i.e. the tokens should still be bonded) or in reality decreasing `usedAllocations` has practically no effect, cause it would be that `usedAllocations > totalAllocations` already anyway.

Given that we have decided to keep this module lean and as independend as possible, I propose that we only make sure that `msgParticipate` checks whether the `auction.StartTime < block.Height + stakingKeeper.UnbondingPeriod(ctx)` and prevents participation otherwise. This is also in line with other launchpads where you can't really "pre-participate" and you can only take action when the auction starts. In our model we allow instead pre-subscription to an auction while it is in standby, so we need to account for that.
I am assuming we don't want to do things like preventing unbonding tokens if it would make `usedAllocations > totalAllocations`, as this would mean requiring to add hooks into staking (if this specifically is even possible), so we have to assume people can (and will) unstake right after sending `msgParticipate`. So the only other option I really see to prevent gamification is the one above. Once an auction has started, you can't participate anymore even with other accounts so there's no "double-spending" possibility.

Moreover, I also propose users can actually claim back allocations right after the auction has *started*, instead of *ended*. As a matter of fact, with our current model we can't prevent a user from

1. bonding tokens
2. Participating to an auction
3. unstaking right after, sending to another account so funds are "immediately" usable (after the unbonding period)

This also means there's no point in enforcing allocations withdrawal *after* the auction has ended, as it would only make the user-experience worse for honest users (assuming we have made sure that the user participated no earlier than `auction.StartTime - stakingKeeper.UnbondingPeriod(ctx)`).

On top of that, I think if you kept your tokens bonded until the auction has started and you can now take part to it you should now be able to do whatever you want with your funds (and therefore, your allocations) because you fulfilled your obligation. But this is more on a "philosophical" level

Opinions? @lubtd @aljo242 @Pantani

**Comment by giunatale** [comment]
decision on this blocks #576 (and potentially #575)

**Comment by giunatale** [comment]
I like the idea. As a matter of fact I hadn't though about that, good point.
Do you think we would still make this ratio of unbonding time a module param?
Like

```
AllocationsUsabilityWindow = {
   ParticipationUnbondingTimeRatio
   WithdrawalUnbodingTimeRatio
}
```
Or rather we remove the param entirely and use constants?

**Comment by lubtd** [comment]
Using params make sense but representing ratios would be too complex in my opinion

I would suggest having two params
```
RegistrationPeriod int64 - delay before auction start, MsgParticipate can be called
WithdrawalDelay int64 ( again :) ) - delay after the auction start when the allocations can be withdrawn
```

IMO we don't even have to link `UnbondingPeriod` in this module, we just need to set the params accordingly

**Comment by lubtd** [comment]
Good, don't hesitate to create new issues now if this will be handled in separate PRs

 I think we can already start creating a `MsgWithdrawAllocations`. For full implementation let's wait for @Pantani and @aljo242 opinions
